src/controller/artifact/annotation/checker/annotation.go

```diff
@@ -0,0 +1,86 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"fmt"
+
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/pkg/artifact"
+)
+
+const (
+	// XIcon is icon content type image/x-icon
+	XIcon = "image/x-icon"
+	// GIF is icon content type image/gif
+	GIF = "image/gif"
+	// PNG is icon content type image/png
+	PNG = "image/png"
+	// JPEG is icon content type image/jpeg
+	JPEG = "image/jpeg"
+
+	// AnnotationPrefix is the prefix of annotation
+	AnnotationPrefix = "io.goharbor.artifact"
+
+	// SkipList is the key word of skip-list annotation
+	SkipList = "skip-list"
+	// Icon is the key word of icon annotation
+	Icon = "icon"
+)
+
+var (
+	// Registry for registered annotation checkers
+	Registry = map[string]AnnotationChecker{}
+
+	// SortedAnnotationVersionList define the order of AnnotationChecker from high to low version.
+	// High version annotation checker will process annotation first.
+	// In the future we have v1alpha2 version, SortedAnnotationVersionList can be defined like
+	// SortedAnnotationVersionList = []string{V1alpha2, V1alpha1}
+	SortedAnnotationVersionList = []string{V1alpha1}
+)
+
+func init() {
+	if err := RegisterV1Alpha1Checker(); err != nil {
+		log.Errorf("Register v1alpha1 annotation Checker error: %v", err)
+	}
+
+	// if err := RegisterV1Alpha2Checker(); err != nil {
+	//	  log.Errorf("Register v1alpha2 annotation checker error: %v", err)
+	// }
+}
+
+// AnnotationChecker process specified annotation
+type AnnotationChecker interface {
```

Comment:
Defining each annotation as a method isn't extensible. When we add a new annotation in the v1alpha2, we must add a new method to the interface, that means we must change the implementation of v1alpha1?

Reply:
How about defining a parse of annotation as following?
```golang
type Parser interface{
    Parse(ctx context.Context, artifact *artifact.Artifact, manifest []byte)(err **error)**
}
```
One parser is responsible for processing all of the annotations belong to one scheme version:
- Parse the annotation from the manifest
- Process the metadata of the artifact according to the annotations
- Store the metadata into the provided artifact model

And there is another parser wrapper which also implements the `Parser` interface, it calls the registered parsers one by one to cover all versions of annotation

Reply:
Yes, the extensibility is not so good. But it is a clear way for different version of annotation checker. And higher version checker will process annotation first.

Reply:
Thought about defined each annotation as interface. But different version annotation checker can not be a same type.

Reply:
> How about defining a parse of annotation as following?
> 
> ```go
> type Parser interface{
>     Parse(ctx context.Context, artifact *artifact.Artifact, manifest []byte)(err **error)**
> }
> ```
> 
> One parser is responsible for processing all of the annotations belong to one scheme version:
> 
> * Parse the annotation from the manifest
> * Process the metadata of the artifact according to the annotations
> * Store the metadata into the provided artifact model
> 
> And there is another parser wrapper which also implements the `Parser` interface, it calls the registered parsers one by one to cover all versions of annotation

I think I understand your idea, we can discussed it later about the detail.